keps/sig-scheduling/5007-device-attach-before-pod-scheduled/README.md

```diff
@@ -0,0 +1,1030 @@
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### DRA Scheduler Plugin Design Overview
+
+This document outlines the design of the DRA Scheduler Plugin, focusing on the handling of fabric devices.
+Key additions include new attributes for device identification, enhancements to `AllocatedDeviceStatus`, and the process for handling `ResourceSlices` upon attachment failure.
+The composable controller design is also discussed, emphasizing efficient utilization of fabric devices.
+
+//TODO update figure
+![proposal](proposal2.JPG)
+
+#### Device Attribute Additions
+
+To indicate whether a device is a fabric device, an attribute is added to the `Basic` within `Device`.
+This attribute will be used by the controller that exposes the `ResourceSlice` to notify whether the device is a fabric device.
+To avoid impacting existing DRA functionality, the default value of this attribute is set to `false`.
+
+```go
+// Device represents one individual hardware instance that can be selected based
+// on its attributes. Besides the name, exactly one field must be set.
+type Device struct {
+    // Name is unique identifier among all devices managed by
+    // the driver in the pool. It must be a DNS label.
+    //
+    // +required
+    Name string
+
+    // Basic defines one device instance.
+    //
+    // +optional
+    // +oneOf=deviceType
+    Basic *BasicDevice
+}
+
+// BasicDevice represents a basic device instance.
+type BasicDevice struct {
+    // Attributes contains additional attributes of the device.
+    //
+    // +optional
+    Attributes map[string]string
+}
+```
+
+To indicate a fabric device, the following attribute will be added:
+
+```yaml
+attributes:
+  kubernetes.io/needs-attaching:
+    bool: "true"
+```
+
+#### `AllocatedDeviceStatus` Additions
+
+The `Conditions` field within `AllocatedDeviceStatus` is used to indicate the status of the device attachment.
+This field will contain a list of conditions, each representing a specific state or event related to the device.
+
+For this feature, the NodeName and following `ConditionType` constants are added:
+
+```go
+// AllocatedDeviceStatus contains the status of an allocated device, if the
+// driver chooses to report it. This may include driver-specific information.
+type AllocatedDeviceStatus struct {
+...
+    // NodeName contains the name of the node where the device needs to be attached.
+    //
+    // +optional
+    NodeName string 
+}
+const(
```

Comment:
we don't need any of these pre-defined in k8s, if we go with the binding gates approach. These all become vendor specific content published by the driver.

The DRA machinery just needs to:
- look for the gates, if they are there then look at the conditions and wait until they are met
- look for the failure conditions and check for failure while waiting
- have some timeout value (I think it should also be in the ResourceSlice)

Reply:
As you point out, a conditional approach is preferable for DRA, so I will update the KEP to reflect that we are using a binding gate approach.

Reply:
It seems like it needs some more discussion, so I'll update the KEP later.